### PR TITLE
[BE][Torch 2.12] Remove workaround code for fixed cublas issue

### LINKED ISSUE #181248: [vllm] [2.12 regression][B200] test_batch_invariance: nondeterministic outputs 3/5 trials with FLASH_ATTN (B200 only, H100 passes)
## Summary

Under torch 2.12.0 + triton 3.7.0, vLLM's `test_v1_generation_is_deterministic_across_batch_sizes_with_needle[FLASH_ATTN]` fails on **B200** because outputs diverge across batch sizes:

> `Failed: Nondeterministic outputs detected: 3 failed out of 5 trials (max_batch_size=128).`

Job reports: `[determinism] total=5, passed=2, failed=3, max_batch_size=128`

The test generates the same prompts under different batch sizes and asserts output identity; a single test run produces identical outputs, but across batch sizes the outputs diverge. Note: **H100 passes** the same test — only **B200** fails. Passes on torch 2.11. Blocking the torch 2.12 upgrade for vLLM (vllm-project/vllm#40077).

## Environment

- `torch`: 2.12.0+cu130 (test channel)
- `triton`: 3.7.0
- CUDA: 13.0
- Python: 3.12.13
- GPU: NVIDIA B200 (fails) — NVIDIA H100 (passes the same test)
- Attention backend: `FLASH_ATTN`

## Reproduction

Failing test:
```
tests/v1/determinism/test_batch_invariance.py::test_v1_generation_is_deterministic_across_batch_sizes_with_needle[FLASH_ATTN]
```

Repro:
```
export VLLM_TEST_MODEL=Qwen/Qwen3-30B-A3B-Thinking-2507-FP8
pytest -v -s 'v1/determinism/test_batch_invariance.py::test_v1_generation_is_deterministic_across_batch_sizes_with_needle[FLASH_ATTN]'
```

The test runs 5 trials; each trial compares a single-request generation to the same prompt batched with other requests. 3 of 5 trials produced mismatching outputs on B200.

## Sample divergence

```
== Not the same as ==

100-story building. A man was standing on the 100th floor. He was holding a glass of water.
He dropped the glass, and it fell 100 floors, yet it didn't break. How is this possible?
```

(full single-vs-batched output pairs captured in job log)

## Reproducibility

- torch-2.12 branch (B200) — still reproducing on every test-PR run since 2026-04-22:
  - 2026-04-22: https://buildkite.com/vllm/ci/builds/62583#019db793-6c6b-4e37-9e80-a0663a043dfa
  - 2026-04-24: https://buildkite.com/vllm/ci/builds/62848#019dbf56-e7f6-4e63-9feb-e137a1b65df7
  - 2026-04-27: https://buildkite.com/vllm/ci/builds/63095#019dcf15-802a-456a-aae7-93c4931e983e
  - 2026-04-28: https://buildkite.com/vllm/ci/builds/63283#019dd3d1-32bb-4326-b3b4-e03529d65006
  - 2026-04-29: https://buildkite.com/vllm/ci/builds/63658#019dda93-ef28-45fe-996b-cd49d3524638
  - All five runs hit the identical signature on the same parametrization: `[determinism] total=5, passed=2, failed=3, max_batch_size=128` on `test_v1_generation_is_deterministic_across_batch_sizes_with_needle[FLASH_ATTN]`
- torch-2.12 branch (H100): ✅ passes (same commit; H100 always passes)
- main (torch 2.11) — passes on every recent B200 run:
  - 2026-04-21 daily: https://buildkite.com/vllm/ci/builds/62383
  - 2026-04-22 nightly: https://buildkite.com/vllm/ci/builds/62456
  - 2026-04-22 daily: https://buildkite.com/vllm/ci/builds/62566
  - 2026-04-23 nightly: https://buildkite.com/vllm/ci/builds/62613
  - 2026-04-24 nightly: https://buildkite.com/vllm/ci/builds/62812
  - 2026-04-29 nightly: https://buildkite.com/vllm/ci/builds/63497
  - 2026-04-29 daily: https://buildkite.com/vllm/ci/builds/63685

Prior torch-2.12 builds (62138/62232/62495) had this test blocked earlier in init by pytorch/pytorch#180905 (bmm kernel registration). Now that #180905 is fixed and closed, the test runs to completion and exposes this B200-specific non-determinism.

## Diagnosis request

The divergence is B200-only, which points to an SM-10/Blackwell-specific kernel (flash-attention variant, softmax, matmul, or rotary) producing batch-dependent numerical drift introduced by torch 2.12 or triton 3.7. Could a maintainer investigate which op path differs between B200 and H100 in this regression?

## Links

- vLLM PR: vllm-project/vllm#40077
- Umbrella: pytorch/pytorch#180899

cc @seemethere @malfet @tinglvv @nWEIdia

---

## Update 2026-04-30 (corrected): MIG hypothesis retracted; isolated to specific B200 hardware/driver

An earlier update on this issue claimed the failure was MIG-on-B200-specific. That was wrong — I conflated the MIG-sliced agent of an unrelated job (`Multi-Modal Processor` on `queue=h200_18gb`) with the agent for this `Batch Invariance (B200)` job. **Both the failing B200 agent and the passing H100 agent run on full GPUs with MIG disabled.**

### Verified agent details (from build 63658 nvidia-smi output)

**Failing CI agent** `dgxB200-09-3` (queue=B200):
```
NVIDIA-SMI 580.65.06   Driver 580.65.06   CUDA 13.0
NVIDIA B200, 183359 MiB, MIG M.: Disabled    (full GPU, both GPU 0 and GPU 1)
```

**Passing CI agent** `mithril-h100-pool`:
```
NVIDIA-SMI 580.95.05   Driver 580.95.05   CUDA 13.0
NVIDIA H100 80GB HBM3, 81559 MiB, MIG M.: Disabled    (full GPU)
```

### Reproduction on a full B200 dev box (also MIG disabled)

```
NVIDIA-SMI 595.58.03   Driver 595.58.03   CUDA 13.2
NVIDIA B200, 183359 MiB, MIG M.: Disabled    (full GPU)
```

Same vLLM commit (`1e8d2aca9` from build 63658), same torch wheel (`torch==2.12.0+cu130`, test channel 2026-04-28 23:18 RC), same `triton==3.7.0`, same `[FLASH_ATTN]` parametrization.

| Field | CI `dgxB200-09-3` (❌ fails) | Dev box mine (✅ passes) | CI `mithril-h100-pool` (✅ passes) |
|---|---|---|---|
| System | **NVIDIA DGX B200** | **AWS EC2 `p6-b200.48xlarge`** | k8s pool with H100 |
| GPU model | NVIDIA B200 | NVIDIA B200 | NVIDIA H100 80GB HBM3 |
| GPU memory | 183 GiB | 183 GiB | 80 GiB |
| MIG | Disabled (full GPU) | Disabled (full GPU) | Disabled (full GPU) |
| Driver | **580.65.06** | **595.58.03** | 580.95.05 |
| CUDA (host) | 13.0 | 13.2 | 13.0 |
| GPU UUID | (not captured in default `nvidia-smi` log) | `GPU-eff083f7-0e58-04bf-0f2d-a26dcef6669b` | (not captured) |
| GPU VBIOS | (not captured) | `97.00.C5.00.2D` | (not captured) |
| Result | `passed=2, failed=3` deterministically (5 consecutive runs Apr 22–29) | `passed=5, failed=0` (6 consecutive runs Apr 30 — 3× at `mem_util=0.5`, 3× at `mem_util=0.9`) | passes every run |

So the failure is **not** caused by MIG, **not** by SM count, **not** by memory pressure (`gpu_memory_utilization=0.5` and `0.9` both clean on dev box), and **not** by torch/triton/vLLM versions (identical across all three environments).

### What's actually different

1. **Platform**: CI runs on an **NVIDIA DGX B200** reference system (`dgxB200-09-3`). My passing dev box is an **AWS EC2 `p6-b200.48xlarge`**. These are different OEM platforms with different VBIOS, NVSwitch/NVLink fabric topology, NUMA layout, and PCIe lane allocation — even though the Blackwell die itself is the same.
2. **Host driver version**: CI is on **580.65.06 / CUDA 13.0**, dev box is on **595.58.03 / CUDA 13.2**. The 595 series is roughly 6 months newer; it may include kernel-launch-ordering, UVM, or stream-synchronization fixes that affect the floating-point reduction order in FlashAttention v2.
3. **Specific hardware unit**: even within DGX B200, `dgxB200-09-3` is a single physical node. Possible per-node variation: VBIOS revision (CI's VBIOS isn't captured by the default `nvidia-smi` block in the log), ECC retired pages, persistent SM clock policy, etc.

### Diagnosis request (corrected)

This now looks like an **environment-specific regression** tied to the DGX B200 platform and/or driver 580.65.06 — the same vLLM/torch wheel produces deterministic outputs on AWS p6-b200 with driver 595.58.03. Could a maintainer:

1. Confirm whether driver `580.65.06` has any known FlashAttention v2 / Blackwell numerical issues that were fixed in the `595.x` series.
2. Reproduce on **another DGX B200** (different physical unit) — if every dgxB200 reproduces, it's a driver/firmware issue. If only `dgxB200-09-3` reproduces, it's per-unit hardware state.
3. **Capture VBIOS/UUID in CI** by adding `nvidia-smi --query-gpu=vbios_version,uuid,serial --format=csv` to the test setup script, so we can compare exactly which DGX firmware vs which AWS firmware behaves how.

A targeted repro inside `dgxB200-09-3` with driver upgraded to 595.x (matching the AWS dev box) would tell us whether the driver alone is the culprit.

### PR DESCRIPTION
**Pull request by Lucaskabela**
## Purpose

Fixes https://github.com/pytorch/pytorch/issues/181248

To do so, we modify vllm/model_executor/layers/batch_invariant.py:enable_batch_invariant_mode() — restrict the Triton persistent matmul override (mm/addmm/matmul/linear) to SM80 (Ampere) only. Hopper and Blackwell now share the same path                                    

The SM100 branch was added as a workaround for a torch-2.9-era cuBLAS bug where B200 selected a non-batch-invariant GEMV path at bs=1 ([pytorch/pytorch#166735](https://github.com/pytorch/pytorch/pull/166735)). The comment in the original code names that PR explicitly. With torch 2.12, cuBLASLt on B200 is batch-invariant at bs=1, so the workaround is no longer load-bearing — and keeping it forces every linear layer through a Triton kernel that is unique to B200, creating an SM100-only divergence surface against the platforms that pass the test.                                                                                                                                                                                  
                                                                  
After the patch, B200 runs exactly the same code path as H100, which has been passing the determinism suite consistently across recent torch releases.
   
## Test Plan
All run on B200 + torch 2.12.0+cu130 + triton 3.7.0 + vllm@HEAD with the patch.                                                                                                                                                    

```bash                                                  
VLLM_TEST_SEED=12345 VLLM_NEEDLE_TRIALS=10 pytest -v -s tests/v1/determinism/test_batch_invariance.py::test_v1_generation_is_deterministic_across_batch_sizes_with_needle -k FLASH_ATTN                                                                                     
```
  Caveats
  - We never reproduced the CI failure (50/50 trials passed pre-patch on FLASH_ATTN), so the local pass tells us only that the patch doesn't break the working case. The decisive signal is CI: needs to be run against a B200 runner that previously failed.
```bash
pytest tests/v1/determinism/test_batch_invariance.py
```

## Test Result
```
Result: [determinism] total=10, passed=10, failed=0, max_batch_size=128.                                                                                                                                                           
```

```bash
tests/v1/determinism/test_batch_invariance.py .........                                                                                                                                                                                        [100%]

================================================================================================================== warnings summary ==================================================================================================================
<frozen importlib._bootstrap>:241
  <frozen importlib._bootstrap>:241: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:241
  <frozen importlib._bootstrap>:241: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

tests/v1/determinism/test_batch_invariance.py:25
  /home/lucaskabela/vllm/tests/v1/determinism/test_batch_invariance.py:25: PytestUnknownMarkWarning: Unknown pytest.mark.timeout - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.timeout(1000)

tests/v1/determinism/test_batch_invariance.py: 14 warnings
  /home/lucaskabela/pytorch/torch/jit/_script.py:365: DeprecationWarning: `torch.jit.script_method` is deprecated. Please switch to `torch.compile` or `torch.export`.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
===================================================================================================== 9 passed, 17 warnings in 316.77s (0:05:16) =====================================================================================================
sys:1: DeprecationWarning: builtin type swigvarlink has no __module__ attribute
(pytorch_build) [lucaskabela@devgpu017.snb3 ~/vllm (main)]$ 
```
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by Lucaskabela** [comment]
cc @zou3519 @BoyuanFeng

**Comment by Lucaskabela** [comment]
Todo - decide if 1) This is torch.compile 2) Check with Paul on the original logic

**Review by yewentao256 — COMMENTED** [review]
Thanks for the work! Could you verify with other unit tests in the test file?

**Comment by Lucaskabela** [comment]
> Todo - decide if 1) This is torch.compile 2) Check with Paul on the original logic

This indeed is related to torch.compile - we use torch.compile on these tests for SM >= 80
```bash
ngineCore pid=2973109) INFO 04-27 13:23:01 [backends.py:1069] Using cache directory: /home/lucaskabela/.cache/vllm/torch_compile_cache/6f63eba5b2/rank_0_0/backbone for vLLM's torch.compile
(EngineCore pid=2973109) INFO 04-27 13:23:01 [backends.py:1128] Dynamo bytecode transform time: 6.05 s
(EngineCore pid=2973109) INFO 04-27 13:23:14 [backends.py:376] Cache the graph of compile range (1, 16384) for later use
(EngineCore pid=2973109) INFO 04-27 13:23:41 [backends.py:391] Compiling a graph for compile range (1, 16384) takes 40.01 s
(EngineCore pid=2973109) INFO 04-27 13:23:44 [monitor.py:53] torch.compile and initial profiling/warmup run together took 49.18 s in total
```

**Comment by Lucaskabela** [comment]
> Thanks for the work! Could you verify with other unit tests in the test file?

@yewentao256 Updated summary to show results of running all tests in this file! Seems to be passing

File: vllm/model_executor/layers/batch_invariant.py [hunk 1]
```diff
@@ -930,29 +930,32 @@ def enable_batch_invariant_mode():
...
         _original_cublas_workspace_cfg = os.environ.get("CUBLAS_WORKSPACE_CONFIG", None)
         _original_cublaslt_workspace_size = os.environ.get(
             "CUBLASLT_WORKSPACE_SIZE", None
         )
         os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":16:8"
         os.environ["CUBLASLT_WORKSPACE_SIZE"] = "1"
 
+    # Triton bmm/persistent-matmul kernels read this for the FP16 N-tile size;
+    # set unconditionally because bmm is overridden on all CUDA platforms.
+    _fp16_block_size_n = 256 if get_max_shared_memory_bytes() > 106496 else 128
```
Comment (yewentao256) [inline vllm/model_executor/layers/batch_invariant.py]: `get_max_shared_memory_bytes` is only for cuda, but currently seems CPU worker may go for this path?

Comment (Lucaskabela) [reply vllm/model_executor/layers/batch_invariant.py]: Ah sorry this leaked out let me revert to back to gate behind CUDA capability

File: vllm/model_executor/layers/batch_invariant.py [hunk 2]
```diff
@@ -930,29 +930,33 @@ def enable_batch_invariant_mode():
...
     else:
-        # Only source of batch invariance for Hopper is split-k, can disable through
-        # cuBLAS workspace config
+        # Hopper (SM90) and Blackwell (SM100): the only source of batch
+        # variance is split-k, which we disable via the cuBLAS workspace
+        # config. The previous SM100 override that routed mm/addmm/matmul/
+        # linear through the triton persistent kernel was a workaround for a
+        # torch 2.9 cuBLAS GEMV-at-bs=1 path (pytorch/pytorch#166735); torch
+        # 2.12 makes cuBLASLt batch-invariant on B200 directly, so B200 now
+        # uses the same path as H100.
```
Comment (yewentao256) [inline vllm/model_executor/layers/batch_invariant.py]:
```suggestion
        # Hopper (SM90) and Blackwell (SM100): the only source of batch
        # variance is split-k, which we disable via the cuBLAS workspace
        # config.
```
We don't need to mention previous update

**Review by yewentao256 — APPROVED** [review]
LGTM, thanks for the work!

Only one small update

**Comment by Lucaskabela** [comment]
Validated this test is failing on main, unrelated to this PR

**Review by atalman — APPROVED** [review]
